AuthProxy: Optimistic lock pattern for remote cache Set

papagian opened the pull request:
Fixes #17375

The error can be triggered by enabling authentication proxy, setting `ldap_sync_ttl` to something small and running the load tests using MySQL or PostgreSQL as a database.

The error occurs when one request tries to insert the key to the table after having checked that it does not exist but in the meantime another one has already inserted it.

The solution attempts to insert the key without checking if it's already there and in case of an integrity constrain violation tries to update it instead.

The transaction is removed in order to prevent deadlocks after an integrity constrain violation occurs. However, there were still some deadlocks on MySQL therefore in case of a deadlock an update is attempted.

 If the update also fails the error will be propagated (since it's a cache and it does not harm a lot to end up with a key that is not set) unless it is a deadlock. In that case the error is ignored since most probably somebody else is trying to insert or update the key, so one way or another the key will be set.

papagian committed:
Implementation of optimistic lock pattern

Try to insert the remote cache key and handle integrity error if occurs

papagian committed:
Remove transaction

Integrity error inside a transaction results in deadlock

papagian committed:
Remove check for existing remote cache key

Is no longer needed since integrity constrain violations are handled

papagian committed:
Add check for integrity constrain violation

Do not update row if the insert statement fails
for other than integrity constrain violation

papagian committed:
Handle failing inserts because of deadlocks

If the insert statement fails because of a deadlock
try to update the row.

papagian committed:
Add utility function for returning SQL error code

Useful for debugging

papagian committed:
Add logging for failing expired cache key deletion

Do not shallow it completely

bergquist commented on pkg/services/sqlstore/migrator/dialect.go:
```diff
@@ -48,6 +48,8 @@ type Dialect interface {
 
 	IsUniqueConstraintViolation(err error) bool
+	IsDeadlock(err error) bool
+	GetErrorCode(err error) string
```
I can't find any usage of this function. Was it added before `IsDeadlock` and not needed anymore?

papagian replied:
It is not needed for this specific case, I included it just because I thought it could be useful being there for the investigation of similar future issues.

bergquist replied:
Since different datasources have different error codes its not really usable beyond testing one database. So I dont think we need it. The dialect function you added is a better suggestion for how people can solve this in the future :)

aocenas replied:
I would say as `GetErrorCode` and `isThisError` are very similar but `GetErrorCode` is probably easier to understand from the name. So what about we keep `GetErrorCode`, remove the `isThisError` and use:
```
func (db *Mysql) IsUniqueConstraintViolation(err error) bool {
	return db.GetErrorCode(err) == mysqlerr.ER_DUP_ENTRY
}
```

papagian replied:
Since the errors for each database are of different type the `GetErrorCode ` returns string.
Therefore this condition would need casting but anyway I think it's better to get rid of `GetErrorCode` and maybe find an more appropriate name for `isThisError`.

bergquist commented on pkg/infra/remotecache/database_storage.go:
```diff
@@ -70,7 +70,10 @@ func (dc *databaseCache) Get(key string) (interface{}, error) {
-			_ = dc.Delete(key) //ignore this error since we will return `ErrCacheItemNotFound` anyway
+			err = dc.Delete(key) //ignore this error since we will return `ErrCacheItemNotFound` anyway
+			if err != nil {
+				dc.log.Debug(err.Error())
```
We should write a log message here as well. Its gonna be very hard to understand why we log this error :)

papagian replied:
I will fix that.

bergquist reviewed (commented):
Great work!

papagian committed:
Revert "Add utility function for returning SQL error code"

This reverts commit 8e0b82c79633e7d8bc350823cbbab2ac7a58c0a5.

aocenas commented on pkg/infra/remotecache/database_storage.go:
```diff
@@ -84,35 +87,37 @@ func (dc *databaseCache) Get(key string) (interface{}, error) {
+	sql := `INSERT INTO cache_data (cache_key,data,created_at,expires) VALUES(?,?,?,?)`
+	_, err = session.Exec(sql, key, data, getTime().Unix(), expiresInSeconds)
+	if err != nil {
+		if dc.SQLStore.Dialect.IsUniqueConstraintViolation(err) || dc.SQLStore.Dialect.IsDeadlock(err) {
```
I think it would be nice to leave some comments in this function as it wasn't very straightforward to get to this point. Things I think should be mentioned here or in function doc comment:
- The deadlock check, that it is only because issues on Mysql and that in this context it should be fine
- That we are doing just one insert/update check and that we can end up with value not being set but as this is cache it is ok (but not good to copy to other context probably)

papagian committed:
Better log for failing deletion of expired cache key

aocenas commented:
@papagian good work 👍. Please add also a bit of description to the PR. No need to be verbose but for other people checking the PR it is good place to leave a bit of context, like what was done and why especially with PRs like this where quite a few decisions and different approaches were made in the process.

bergquist commented on pkg/infra/remotecache/database_storage.go:
```diff
@@ -84,35 +87,37 @@ func (dc *databaseCache) Get(key string) (interface{}, error) {
-			expiresInSeconds = int64(expire) / int64(time.Second)
-		}
+	var cacheHit CacheData
+	_, err = session.Where("cache_key = ?", key).Get(&cacheHit)
```
Good point! But I think we should avoid the insert statement if this statement returns something. Should reduce the amounts of locks in the database.

aocenas replied:
I think that would get us back to where we started though, wouldn't it. I mean whatever this returns we cannot count on it when doing the insert/update as other operation could have inserted/deleted the row and we can still get unique error or not update anything.

bergquist replied:
Yea I forgot the delete aspect. I think we can improve this but it should be fine for the only use-case we have now.

bergquist commented:
cant respond to this comment for some reason...
> I would say as GetErrorCode and isThisError are very similar but GetErrorCode is probably easier to understand from the name. So what about we keep GetErrorCode, remove the isThisError and use:

I agree that `GetErrorCode` is better but I dont think we need to expose it in the interface.

papagian committed:
Add some comments

papagian committed:
Remove check for existing cache key

Attempts to insert the key without checking if it's already there
and handle the error situations

papagian committed:
Do not propagate deadlocks created during update

Most probably somebody else is trying to insert/update
the key at the same time so it is safe enough to ignore it
